File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    int shipWithinDays(vector<int>& weights, int days) {
        int limit = 0;
        for(auto x: weights){
            limit += x;
        }
        
        int s = 0;
        int e = limit;
        
        int ans = limit;
        
        while(s <= e){
            // abhi mid maximum weight hai ship ka
            int mid = (s + e) / 2;
            cout<<mid<<" ";
            int i = 0;
            int j = 0;
            int s_l = mid;
            for(i = 1; i <= days; i++){
                while(1){
                    if(j >= weights.size()){
                        break;
                    }
                    if(s_l - weights[j] >= 0){
                        s_l -= weights[j];
                        j++;
                    }else{
                        s_l = mid;
                        break;
                    }
                }
            }
            cout<<i<<" "<<j<<endl;
            //if(mid < 0){
              //  break;
            //}
            if(j >= weights.size()){
                ans = mid;
                e = mid - 1;    
            }else{
                s = mid + 1;       
            }
        }
        return ans;
    }
};

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    private int ans = 0;

    private boolean filledProperly(int maxCapacityPerDay, int[] weights, int days){
        //System.out.println(maxCapacityPerDay);
        int currentCapacityPerDay = 0;

        int i = 0;
        while(i < weights.length){
            if(currentCapacityPerDay + weights[i] > maxCapacityPerDay){
                // can't be filled in current day
                days -= 1;
                currentCapacityPerDay = 0;
            } 

            currentCapacityPerDay += weights[i];

            if(currentCapacityPerDay > maxCapacityPerDay){
                return false;
            }
            i += 1;
        }

        if(days <= 0){
            return false;
        }

        return true;
    }

    private void helper(int left, int right, int[] weights, int days){
        if(left > right){
            return;
        }

        int mid = (left + right) / 2;

        if(filledProperly(mid, weights, days)){
            ans = mid;
            helper(left, mid - 1, weights, days);
        }else{
            helper(mid + 1, right, weights, days);
        }
    } 

    public int shipWithinDays(int[] weights, int days) {
        int left = 0, right = Integer.MAX_VALUE;
        helper(left, right, weights, days);
        return ans;
    }
}